### PR TITLE
Replace handwritten SIMD implementation with autovectorization for +10% perf

### PR DESCRIPTION
**Pull request by Shnatsel**
This gives a +10% performance boost on 3bbp and 6bbp unfiltering benchmarks. 

cc @anforowicz who wrote the explicit SIMD code

The downside of autovectorization is that it can be finicky and is not at all guarateed. So this may potentially be more fragile than explicit SIMD.

On the flip side, we've been relying on autovectorization for performance for ages and it has always worked once we got it working once. I've checked [this sample](https://godbolt.org/z/8jTEr1Kbv) on Godbolt and it didn't vectorize on Rust v1.60, but every single version starting from 1.61 does vectorize it.

ARM seems to benefit as well, resulting in vectorized and shorter code on godbolt. I don't have an aarch64 machine to measure it though.

### CODE REVIEW

**Comment by Shnatsel** [comment]
There's some background on this in #511

I'm still hoping to coax autovectorization into working without any `std::simd` usage at all, but I haven't managed it so far. If I succeed, that will probably supersede this PR - but I figured it's a good idea to start the discussion about replacing `std::simd` with autovectorization early.

**Comment by fintelia** [comment]
> figured it's a good idea to start the discussion about replacing std::simd with autovectorization early, in case anyone is opposed to this approach.

I support the switch. Using `std::simd` was only begrudgingly accepted when we couldn't figure out a way to get the relevant code to autovectorize.

**Comment by Shnatsel** [comment]
To clarify, this PR still requires `std::simd` types in function signatures, and thus still requires nightly compiler.

I'm still trying (and failing) to get it to work on stable. For the purposes of this review let's assume that nightly is still going to be required, even for the autovec implementation.

**Comment by Shnatsel** [comment]
Okay I've finally figured out the 3 and 6 bpp case, PR up: #513

**Comment by Shnatsel** [comment]
This performs better in benchmarks, and produces shorter but fully vectorized assembly on both x86_64 and Aarch64. 

Since this seems to be an unambiguous improvement, I'm going to go ahead and merge this so that the PR doesn't stall.

File: src/filter.rs [hunk 1]
```diff
@@ -32,28 +32,26 @@ mod simd {
...
+    /// [the spec](http://www.libpng.org/pub/png/spec/1.2/PNG-Filters.html#Filter-type-4-Paeth)
+    /// except that it simultaneously calculates the predictor for all SIMD lanes.
+    /// Mapping between parameter names and pixel positions can be found in
+    /// [a diagram here](https://www.w3.org/TR/png/#filter-byte-positions).
+    ///
+    /// Examples of how different pixel types may be represented as multiple SIMD lanes:
+    /// - RGBA => 4 lanes of `i16x4` contain R, G, B, A
+    /// - RGB  => 4 lanes of `i16x4` contain R, G, B, and a ignored 4th value
+    ///
+    /// The SIMD algorithm below is based on [`libpng`](https://github.com/glennrp/libpng/blob/f8e5fa92b0e37ab597616f554bee254157998227/intel/filter_sse2_intrinsics.c#L261-L280).
```
Comment (okaneco) [inline src/filter.rs:52]: This comment isn't correct.

As it states on L72, the logic is a translation of how pa, pb, and pc are calculated in `filter::filter_paeth` https://github.com/image-rs/image-png/blob/3fbbbb188e967547563496282591c52fc5892746/src/filter.rs#L362-L401

Comment (Shnatsel) [reply src/filter.rs:52]: Thanks for pointing this out!

I didn't actually add this comment, it's an artifact of how Github calculated the diff. You can see the same lines removed further up.

I'd be happy to merge a PR correcting it, though!

**Pull request merged**